### PR TITLE
Posts without text or only with emojis are now always accepted in the language check

### PR DESCRIPTION
**Pull request by annando**
When posts are received via a relay server, we can check if they contain some unwanted languages. We now always accept posts without a body (picture posts) or with only emojis in it, since they don't contain the unwanted languages.

### CODE REVIEW

File: src/Content/Smilies.php [hunk 1]
```diff
@@ -285,4 +285,17 @@ private static function heartReplaceCallback(array $matches): string
...
+
+	/**
+	 * Checks if the body only contains 4 byte unicode characters.
+	 *
+	 * @param string $body
+	 * @return boolean
+	 */
+	public static function isEmojiPost(string $body): bool
+	{
+		$conv = html_entity_decode(str_replace([' ', "\n", "\r"], '', $body));
```
Comment (MrPetovan) [inline src/Content/Smilies.php]: I think you should do a more comprehensive `preg_replace('#\s#', '', html_entity_decode($body))` to catch potential non-breaking spaces that can be HTML-encoded. `\s` catches all whitespace characters.

A suite of unit tests would be welcome to catch potential corner cases like this one I thought about.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Content/Smilies.php (hunk 1 above)
Comment (annando) [reply src/Content/Smilies.php]: Done

File: src/Content/Smilies.php [hunk 2]
```diff
@@ -285,4 +285,17 @@ private static function heartReplaceCallback(array $matches): string
...
+
+	/**
+	 * Checks if the body only contains 4 byte unicode characters.
+	 *
+	 * @param string $body
+	 * @return boolean
+	 */
+	public static function isEmojiPost(string $body): bool
+	{
+		$conv = html_entity_decode(preg_replace('#\s#', '', $body));
```
Comment (MrPetovan) [inline src/Content/Smilies.php]: This will still be stunted by `&nbsp;` entities that should be decoded first before running through `preg_replace`.
```suggestion
		$conv = preg_replace('#\s#', '', html_entity_decode$body));
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Content/Smilies.php (hunk 2 above)
Comment (annando) [reply src/Content/Smilies.php]: Done

File: src/Content/Smilies.php [hunk 3]
```diff
@@ -294,7 +294,7 @@ private static function heartReplaceCallback(array $matches): string
 	 */
 	public static function isEmojiPost(string $body): bool
 	{
-		$conv = html_entity_decode(preg_replace('#\s#', '', $body));
+		$conv = preg_replace('#\s#', '', html_entity_decode($body));
```
Comment (MrPetovan) [inline src/Content/Smilies.php]: Just tested with `&nbsp;` and it needs this modifier to catch the UTF-8 non-breaking space character in the subject:
```suggestion
		$conv = preg_replace('#\s#u', '', html_entity_decode($body));
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Comment by MrPetovan** [comment]
I started writing unit tests, and I found out that this emoji doesn't validate the method: 👨‍👨‍👧. This is because it is a composite emoji made of 👨👨 and 👧 (each being 4-byte characters) separated by Zero-Width Joiners characters (encoded on only 3 bytes).

**Review by MrPetovan — CHANGES_REQUESTED** [review]
I started writing unit tests, and I found out that this emoji doesn't validate the method: 👨‍👨‍👧. This is because it is a composite emoji made of 👨👨 and 👧 (each being 4-byte characters) separated by Zero-Width Joiners characters (encoded on only 3 bytes). Back to the drawing board!

**Comment by annando** [comment]
How do we continue with this PR?

**Comment by MrPetovan** [comment]
We have to find another way of detecting emoji-only posts, the 4-byte rule isn't universal and will lead to inconsistent results.

**Comment by annando** [comment]
Yes. But this one is better than nothing. This PR here just moves the existing check into a separate function. And at the places where the function is used, some false negatives don't really hurt.

**Comment by MrPetovan** [comment]
Then we have to make it clear in the method description that it is an approximation, possibly including in the method name, and comment out the failing test cases in order for tests to pass anyway,

**Comment by annando** [comment]
We should just put it into the description. Eventually someone will have a look at the function and will improve it.

**Comment by MrPetovan** [comment]
"Eventually" is load-bearing here, but the unit test suite I added to your branch shouldn't be failing while we wait for a savior, so it should be commented to fit the present decision.

**Comment by tobiasd** [comment]
In case this PR shall be part of the 2023.09 (10) release, please redirect it towards the 2023.09-rc branch.

**Comment by annando** [comment]
I'm now using the functionality from the `IntlChar` class.

**Review by MrPetovan — APPROVED** [review]

**Comment by MrPetovan** [comment]
How do the unit tests look?

**Comment by annando** [comment]
I haven't done anything with the tests, so we have to see.

**Comment by MrPetovan** [comment]
These can be ran directly with PHPUnit without the whole Continuous Integration infrastructure as they don’t require a database connection, I’ll check out your branch and run the tests manually tonight or tomorrow EST.

**Comment by MrPetovan** [comment]
Here are the only two tests cases that are failing:
```
'emojis+HTML else' => [
	'expected' => false,
	'body' => '👀&lt;🤷',
],
'HTML entity else' => [
	'expected' => false,
	'body' => '&deg;',
],
```
`Smilies::isEmojiPost` unexpectedly returns `true` for both cases.

**Comment by annando** [comment]
I currently have slight issues with `°`. This is a two byte graphical character that isn't alphanumeric. One could possibly argue, that it can be used to express emotions as well?

**Comment by MrPetovan** [comment]
Sure, but in this case you have to be more precise about the meaning of `Smilies::isEmojiPost` up to and including renaming it to match the updated meaning.

I wrote these tests based on its current description, the tests can be changed, but it needs to be accompanied by a change in the method itself, either the code and/or the name and/or the documentation.

**Comment by annando** [comment]
I hope that the added check will now work well for all characters.

**Comment by MrPetovan** [comment]
It should, but in the shower I thought about non-Latin scripts like Arabic script, I'm pretty sure it doesn't pass the check.

**Comment by annando** [comment]
The function `IntlChar::isalnum` returns "true" for characters in all scripts of the world. I'm using this in the language detection part as well.

**Review by MrPetovan — APPROVED** [review]
